security/src/main/java/io/airlift/security/cert/CertificateBuilder.java

```diff
@@ -0,0 +1,262 @@
+        return this;
+    }
+
+    public CertificateBuilder setSubject(X500Principal subject)
+    {
+        this.subject = requireNonNull(subject, "subject is null");
+        return this;
+    }
+
+    public CertificateBuilder addSanIpAddress(InetAddress address)
+    {
+        this.sanIpAddresses.add(requireNonNull(address, "address is null"));
+        return this;
+    }
+
+    public CertificateBuilder addSanIpAddresses(List<InetAddress> addresses)
+    {
+        requireNonNull(addresses, "addresses is null");
+        addresses.forEach(this::addSanIpAddress);
+        return this;
+    }
+
+    public CertificateBuilder addSanDnsName(String dnsNames)
+    {
+        this.sanDnsNames.add(requireNonNull(dnsNames, "dnsNames is null"));
+        return this;
+    }
+
+    public CertificateBuilder addSanDnsNamees(List<String> dnsNames)
+    {
+        requireNonNull(dnsNames, "dnsNames is null");
+        dnsNames.forEach(this::addSanDnsName);
+        return this;
+    }
+
+    public X509Certificate buildSelfSigned()
+            throws GeneralSecurityException
+    {
+        requireNonNull(publicKey, "publicKey is null");
+        requireNonNull(privateKey, "privateKey is null");
+        requireNonNull(issuer, "issuer is null");
+        requireNonNull(notBefore, "notBefore is null");
+        requireNonNull(notAfter, "notAfter is null");
+        requireNonNull(subject, "subject is null");
+
+        byte[] publicKeyHash = hashPublicKey();
+
+        List<byte[]> sans = new ArrayList<>();
+        sanDnsNames.stream()
+                .map(address -> encodeContextSpecificTag(2, address.getBytes(UTF_8)))
+                .forEach(sans::add);
+        sanIpAddresses.stream()
+                .map(InetAddress::getAddress)
+                .map(address -> encodeContextSpecificTag(7, address))
+                .forEach(sans::add);
+
+        byte[] rawCertificate = encodeSequence(
+                // version: 2
+                encodeContextSpecificSequence(0, encodeInteger(2)),
+                // serialNumber
+                encodeInteger(serialNumber),
+                // signature kind
+                encodeSequence(
+                        SHA_256_WITH_RSA_ENCRYPTION_OID,
+                        encodeNull()),
+                // issuer
+                issuer.getEncoded(),
+                // validity
+                encodeSequence(
+                        //    SEQUENCE (2 elem)
+                        //      UTCTime 2021-01-01 00:00:00 UTC
+                        //      UTCTime 2022-01-01 00:00:00 UTC
+                        // not before
+                        encodeUtcTime(notBefore),
+                        // not after
+                        encodeUtcTime(notAfter)),
+                // subject
+                subject.getEncoded(),
+                // public key
+                publicKey.getEncoded(),
+                // extensions
+                encodeContextSpecificSequence(3, encodeSequence(
```

Comment:
What is the 3?

Reply:
IiRC it means "this is the extensions section"... just like `0` means "this is the verison section"